### PR TITLE
8308588: Unnecessary synchronized on GTKStyle#ICONS_MAP can be removed

### PR DESCRIPTION
**Pull request by kumarabhi006**
`static final Map<String,GTKStockIcon> ICONS_MAP` is modified only in `static` block. Then [com.sun.java.swing.plaf.gtk.GTKStyle#get](https://github.com/kumarabhi006/jdk/blob/73f7a5f15dbba54a98f3916ff1190520ac07874d/src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKStyle.java#L892) method read from it within `synchronized (ICONS_MAP)` block. As this map is read-only and all its content is initialized in static block we can safely remove synchronized from get call.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308588](https://bugs.openjdk.org/browse/JDK-8308588): Unnecessary synchronized on GTKStyle#ICONS_MAP can be removed (**Enhancement** - P4)

### Reviewers
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - **Reviewer**)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Andrey Turbanov](https://openjdk.org/census#aturbanov) (@turbanoff - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20741/head:pull/20741` \
`$ git checkout pull/20741`

Update a local copy of the PR: \
`$ git checkout pull/20741` \
`$ git pull https://git.openjdk.org/jdk.git pull/20741/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20741`

View PR using the GUI difftool: \
`$ git pr show -t 20741`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20741.diff">https://git.openjdk.org/jdk/pull/20741.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20741#issuecomment-2314749835)

### CODE REVIEW

**Review by DamonGuy — COMMENTED** [review]
I remember simply testing this change when originally triaging this. How did you test this change for stability?

**Review by alisenchung — COMMENTED** [review]
What's the benefit of removing synchronized from the get call? Also were tests run on all platforms that use GTK?

**Comment by kumarabhi006** [comment]
> I remember simply testing this change when originally triaging this. How did you test this change for stability?

Yeah I saw your comment on JBS. I couldn't think of writing any test case for this fix. I ran all the regression tests in CI and there was no failure observed for the proposed changes.

**Comment by kumarabhi006** [comment]
> What's the benefit of removing synchronized from the get call? 

As I mentioned earlier, `ICONS_MAP` are initialized only in **static block** and after that it is just referenced to access the value out of it. Even though `ICONS_MAP` are accessed simultaneously for read operation by multiple threads, it won't lead to any  inconsistent state. So, the use of Synchronized look unnecessary to me and the removal of Synchronized shouldn't affect the object state in multi threaded environment.

> Also were tests run on all platforms that use GTK? 

Basically I ran the regression tests for linux platform which is supposed to have GTK environment.

**Review by TejeshR13 — COMMENTED** [review]
Yes, it looks like redundant though. One slight concern is that, does GTKStockIcon has any role to play with synchronized block? If that doesn't has anything with synchronization then its good to go.

**Comment by kumarabhi006** [comment]
> One slight concern is that, does GTKStockIcon has any role to play with synchronized block?

Sorry, I didn't understand what do you mean.

**Comment by prsadhuk** [comment]
> Basically I ran the regression tests for linux platform which is supposed to have GTK environment. Added the CI link in JBS.

FWIW, Running tests on linux environment will test it on default MetalL&F which will not test this GTK path unless you specify GTK L&F in `swing.defaultlaf` to it

**Comment by TejeshR13** [comment]
> > One slight concern is that, does GTKStockIcon has any role to play with synchronized block?
> 
> Sorry, I didn't understand what do you mean.

I meant, since `ICONS_MAP` stores `GTKStockIcon`  as value, I was wondering does that make any difference or need for the map to be synchronized? I also came across Immutable Maps which are inherently thread safe. If ICONS_MAP are immutable one, then Synchronized block is definitely redundant.

**Comment by kumarabhi006** [comment]
> FWIW, Running tests on linux environment will test it on default MetalL&F which will not test this GTK path unless you specify GTK L&F in `swing.defaultlaf` to it

Ok... I will test with GTK L&F option which I missed earlier.

**Comment by kumarabhi006** [comment]
> I meant, since `ICONS_MAP` stores `GTKStockIcon` as value, I was wondering does that make any difference or need for the map to be synchronized? I also came across Immutable Maps which are inherently thread safe. If ICONS_MAP are immutable one, then Synchronized block is definitely redundant.

`GTKStockIcon` objects are also created inside static block, So, I guess that should be fine.

**Comment by kumarabhi006** [comment]
> FWIW, Running tests on linux environment will test it on default MetalL&F which will not test this GTK path unless you specify GTK L&F in `swing.defaultlaf` to it

@prsadhuk Ran the tests with GTK Look and Feel option in CI and the tests failed with or without fix is same except one test failure `JFrameMenuSerializationTest.java`. This test failed with the exception `Exception: java.lang.NullPointerException: Cannot invoke "javax.swing.plaf.synth.SynthStyle.getColor(javax.swing.plaf.synth.SynthContext, javax.swing.plaf.synth.ColorType)" because "style" is null` when run without the fix.

However, I checked the test result without fix in CI pipeline and the test result shows it passed but the html log file shows `java.lang.NullPointerException: Cannot invoke "javax.swing.plaf.synth.SynthStyle.getColor(javax.swing.plaf.synth.SynthContext, javax.swing.plaf.synth.ColorType)" because "style" is null` under System.err output.

Checked in my local machine and this test fails with the same exception.

I think this is not related with the fix proposed here because it seems the **GTKStyle object itself is null** if it is run with GTK L&F option.

Added the CI link when tests are ran with GTK L&F option in JBS.

**Comment by aivanov-jdk** [comment]
> > What's the benefit of removing synchronized from the get call?
> 
> As I mentioned earlier, `ICONS_MAP` are initialized only in **static block** and after that it is just referenced to access the value out of it. Even though `ICONS_MAP` are accessed simultaneously for read operation by multiple threads, it won't lead to any inconsistent state. So, the use of Synchronized look unnecessary to me and the removal of Synchronized shouldn't affect the object state in multi threaded environment.

I agree with @turbanoff's evaluation in [the bug description](https://bugs.openjdk.org/browse/JDK-8308588#descriptionmodule): <q cite="https://bugs.openjdk.org/browse/JDK-8308588#descriptionmodule">As this map is read-only and _all its content is initialized in `clinit`_, we can safely remove `synchronized` from `get` call</q>.

We may go further and make `ICONS_MAP` unmodifiable map:

```java
        Map<String,GTKStockIcon> iconsMap = new HashMap<>();
        iconsMap.put("FileChooser.cancelIcon", new GTKStockIcon("gtk-cancel", 4));
        iconsMap.put("FileChooser.okIcon",     new GTKStockIcon("gtk-ok",     4));
        iconsMap.put("OptionPane.yesIcon", new GTKStockIcon("gtk-yes", 4));
        iconsMap.put("OptionPane.noIcon", new GTKStockIcon("gtk-no", 4));
        iconsMap.put("OptionPane.cancelIcon", new GTKStockIcon("gtk-cancel", 4));
        iconsMap.put("OptionPane.okIcon", new GTKStockIcon("gtk-ok", 4));
        ICONS_MAP = Collections.unmodifiableMap(iconsMap);
```

This guarantees that `ICONS_MAP` cannot be modified; additionally the map is fully initialised by the time it's assigned to `ICONS_MAP` member.

Alternatively, as IDEA suggests, `Map.of` can be used instead:

```java
        ICONS_MAP = Map.of(
                "FileChooser.cancelIcon", new GTKStockIcon("gtk-cancel", 4),
                "FileChooser.okIcon", new GTKStockIcon("gtk-ok", 4),
                "OptionPane.yesIcon", new GTKStockIcon("gtk-yes", 4),
                "OptionPane.noIcon", new GTKStockIcon("gtk-no", 4),
                "OptionPane.cancelIcon", new GTKStockIcon("gtk-cancel", 4),
                "OptionPane.okIcon", new GTKStockIcon("gtk-ok", 4));
```

**Comment by aivanov-jdk** [comment]
> > > One slight concern is that, does GTKStockIcon has any role to play with synchronized block?
> > 
> > 
> > Sorry, I didn't understand what do you mean.
> 
> I meant, since `ICONS_MAP` stores `GTKStockIcon` as value, I was wondering does that make any difference or need for the map to be synchronized?

In fact, `synchronized` block around `ICONS_MAP.get` had no effect.

To be correct, the value for `ICONS_MAP` had to be assigned inside a `synchronized` block too. Otherwise, it is not thread-safe.

> I also came across Immutable Maps which are inherently thread safe. If ICONS_MAP are immutable one, then Synchronized block is definitely redundant.

Effectively, `ICONS_MAP` is an immutable map. It's initialised in a static initialiser and is never modified afterwards.

As far as I can tell, `GTKStockIcon` objects stored in the map aren't modified either.

Thus, the entire structure is immutable and is thread-safe without additional synchronisation.

File: src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKStyle.java [hunk 1]
```diff
@@ -888,10 +888,7 @@ else if ("CheckBox.iconTextGap".equals(key) ||
         }
 
         // Is it a stock icon ?
-        GTKStockIcon stockIcon = null;
-        synchronized (ICONS_MAP) {
-            stockIcon = ICONS_MAP.get(key);
-        }
+        GTKStockIcon stockIcon = ICONS_MAP.get(key);
 
         if (stockIcon != null) {
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKStyle.java:893]:
```suggestion
        GTKStockIcon stockIcon = ICONS_MAP.get(key);
        if (stockIcon != null) {
```
Maybe remove the blank line? It used to separate a block where `stockIcon` was initialised and the condition. Now it's better to keep these two lines closer.

**Review by aivanov-jdk — COMMENTED** [review]
It looks good to me as it is.

However, I'd go further and use either `Map.of` or `Collection.unmodifiableMap` to emphasise the fact that the map is read-only.

**Comment by kumarabhi006** [comment]
@aivanov-jdk I guess `Maps.of` can't be used here because there are few elements put inside a map after `tk.checkGtkVersion(3, 10, 0)` condition evaluation at [L1193](https://github.com/kumarabhi006/jdk/blob/73f7a5f15dbba54a98f3916ff1190520ac07874d/src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/GTKStyle.java#L1193). Instead used `Collections.unmodifiableMap` to make the `ICONS_MAP` read only.

**Review by aivanov-jdk — APPROVED** [review]

**Comment by aivanov-jdk** [comment]
> @aivanov-jdk I guess `Maps.of` can't be used here because there are few elements put inside a map after `tk.checkGtkVersion(3, 10, 0)` condition…

It's possible to use `Map.of` if you put the condition as a conditional operator to each value. Yet it would make the intention less clear.

**Comment by kumarabhi006** [comment]
> It's possible to use `Map.of` if you put the condition as a conditional operator to each value. Yet it would make the intention less clear.

Yeah that can be done but current implementation look ok to me as well.

**Review by turbanoff — APPROVED** [review]

**Review by TejeshR13 — APPROVED** [review]
Looks good to me.
